lib/queryContextCache.js

```diff
@@ -0,0 +1,391 @@
+        qce.setContext(context);
+      } else {
+
+        // Change in priority
+        const newQCE =
+              new QueryContextElement(id, timestamp, priority, context);
+
+        this.replaceQCE(qce, newQCE);
+      } // new priority
+    } // new data is recent
+    else if (timestamp === qce.getReadTimestamp() && qce.getPriority() !== priority) {
+      // Same read timestamp but change in priority
+      const newQCE = new QueryContextElement(id, timestamp, priority, context);
+      this.replaceQCE(qce, newQCE);
+    }
+  } // id found
+  else {
+
+    // new id
+    if (this.priorityMap.has(priority)) {
+
+      // Same priority with different id
+      const qce = this.priorityMap.get(priority);
+
+      // Replace with new data
+      const newQCE = new QueryContextElement(id, timestamp, priority, context);
+      this.replaceQCE(qce, newQCE);
+    } else {
+
+      // new priority
+      // Add new element in the cache
+      const newQCE = new QueryContextElement(id, timestamp, priority, context);
+      this.addQCE(newQCE);
+    }
+  }
+};
+
+/** Sync the newPriorityMap with the priorityMap at the end of current round of merge */
+QueryContextCache.prototype.syncPriorityMap = function () {
+  Logger.getInstance().debug(
+    `syncPriorityMap called priorityMap size = ${this.priorityMap.size}, newPrioirtyMap size = ${this.newPriorityMap.size}`
+  );
+  this.newPriorityMap.forEach((value,key)=>{
+    this.priorityMap.set(key,value);
+  });
+
+  // clear the newPriorityMap for next round of QCC merge(a round consists of multiple entries)
+  this.newPriorityMap.clear();
+};
+
+/**
+   * After the merge, loop through priority list and make sure cache is at most capacity. Remove all
+   * other elements from the list based on priority.
+   */
+QueryContextCache.prototype.checkCacheCapacity = function () {
+  Logger.getInstance().debug(
+    `checkCacheCapacity() called. treeSet size ${this.treeSet.size} cache capacity ${this.capacity}` );
+
+  // remove elements based on priority
+  while (this.treeSet.size > this.capacity) {
+    const qce = Array.from(this.treeSet).pop();
+    this.removeQCE(qce);
+  }
+  Logger.getInstance().debug(
+    `checkCacheCapacity() returns. treeSet size ${this.treeSet.size} cache capacity ${this.capacity}`,
+  );
+};
+
+/** Clear the cache. */
+QueryContextCache.prototype.clearCache = function () {
+  Logger.getInstance().debug('clearCache() called');
+  this.idMap.clear();
+  this.priorityMap.clear();
+  this.treeSet.clear();
+  Logger.getInstance().debug(`clearCache() returns. Number of entries in cache now ${this.treeSet.size}`,);
+};
+
+QueryContextCache.prototype.getElements = function () {
+  return this.treeSet;
+};
+
+/**
+ * @param data: the QueryContext Object serialized as a JSON format string
+ */
+QueryContextCache.prototype.deserializeQueryContext = function (data) {
+  Logger.getInstance().debug(`deserializeQueryContext() called: data from server: ${JSON.stringify(data)}`);
```

Comment:
should we stringify the data many times?

Reply:
I'm not questioning conversion for the debug message, but line belowe we again call stringify - do we need to do it two times?

Reply:
Fixed.